File: backend/src/main/java/kr/co/programmers/collabond/core/config/SecurityConfig.java
package kr.co.programmers.collabond.core.config;

import kr.co.programmers.collabond.core.auth.CustomAuthenticationEntryPoint;
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2SuccessHandler;
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2UserService;
import kr.co.programmers.collabond.core.auth.jwt.TokenAuthenticationFilter;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsUtils;

import java.util.List;

@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {

    @Value("${custom.cors.allowed-origin}")
    private String allowedOrigin;

    private final OAuth2UserService oAuth2UserService;
    private final OAuth2SuccessHandler oAuth2SuccessHandler;
    private final CustomAuthenticationEntryPoint entryPoint;
    private final TokenAuthenticationFilter tokenAuthenticationFilter;

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        return http
                .cors(cors -> cors.configurationSource(
                        request -> {
                            CorsConfiguration configuration = new CorsConfiguration();

                            configuration.setAllowedOrigins(List.of(allowedOrigin));
                            configuration.setAllowedMethods(List.of("*"));
                            configuration.setAllowCredentials(true);
                            configuration.setAllowedHeaders(List.of("*"));
                            configuration.setMaxAge(3600L);

                            configuration.setExposedHeaders(List.of("Set-Cookie", "Authorization"));
                            return configuration;
                        }
                ))
                .csrf(csrf -> csrf.disable()) // csrf 비활성
                .httpBasic(AbstractHttpConfigurer::disable) // 기본 인증 로그인 비활성
                .formLogin(form -> form.disable()) // 기본 로그인 폼 비활성
                .logout(AbstractHttpConfigurer::disable) // 기본 로그아웃 비활성
                .addFilterAfter(tokenAuthenticationFilter,
                        UsernamePasswordAuthenticationFilter.class)
                .oauth2Login(oauth -> {
                    oauth.userInfoEndpoint(userInfo ->
                                    userInfo.userService(oAuth2UserService))
                            .successHandler(oAuth2SuccessHandler);
                })
                // 임시로 Admin만 권한 검사
                .authorizeHttpRequests(
                        auth -> auth
                                .requestMatchers(CorsUtils::isPreFlightRequest).permitAll()
                                .requestMatchers("/admin/**").hasRole("ADMIN")
                                .requestMatchers(HttpMethod.PATCH,"/api/users/signup").hasRole("TMP")
                                .requestMatchers(HttpMethod.GET, "/api/users/{userId}",
                                    "/api/profiles/**",
                                    "/api/recruitments/**",
                                    "/api/tags/**").permitAll()
                                .requestMatchers("/",
                                        "/error",
                                        "/oauth2/**",
                                        "/api/tokens/refresh").permitAll()
                                .anyRequest().authenticated()
                )
                .exceptionHandling(handler -> handler.authenticationEntryPoint(entryPoint))
                .sessionManagement(session ->
                        session.sessionCreationPolicy(SessionCreationPolicy.STATELESS)) // 세션 사용 X
                .build();
    }
}

File: backend/src/main/java/kr/co/programmers/collabond/core/config/SecurityConfig.kt
package kr.co.programmers.collabond.core.config

import kr.co.programmers.collabond.core.auth.CustomAuthenticationEntryPoint
import kr.co.programmers.collabond.core.auth.jwt.TokenAuthenticationFilter
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2SuccessHandler
import kr.co.programmers.collabond.core.auth.oauth2.OAuth2UserService
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsUtils

@Configuration
@EnableWebSecurity
class SecurityConfig(
    @Value("\${custom.cors.allowed-origin}")
    private val allowedOrigin: String,
    private val oAuth2UserService: OAuth2UserService,
    private val oAuth2SuccessHandler: OAuth2SuccessHandler,
    private val entryPoint: CustomAuthenticationEntryPoint,
    private val tokenAuthenticationFilter: TokenAuthenticationFilter
) {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain = http
        .cors { cors ->
            cors.configurationSource {
                CorsConfiguration().apply {
                    allowedOrigins = listOf(allowedOrigin)
                    allowedMethods = listOf("*")
                    allowCredentials = true
                    allowedHeaders = listOf("*")
                    maxAge = 3600L
                    exposedHeaders = listOf("Set-Cookie", "Authorization")
                }
            }
        }
        .csrf { it.disable() }
        .httpBasic { it.disable() }
        .formLogin { it.disable() }
        .logout { it.disable() }
        .addFilterAfter(
            tokenAuthenticationFilter,
            UsernamePasswordAuthenticationFilter::class.java
        )
        .oauth2Login {
            it.userInfoEndpoint { userInfo ->
                userInfo.userService(oAuth2UserService)
            }
            it.successHandler(oAuth2SuccessHandler)
        }
        .authorizeHttpRequests { auth ->
            auth.requestMatchers(CorsUtils::isPreFlightRequest).permitAll()
            auth.requestMatchers("/admin/**").hasRole("ADMIN")
            auth.requestMatchers(HttpMethod.PATCH,"/api/users/signup").hasRole("TMP")
            auth.requestMatchers(HttpMethod.GET, "/api/users/{userId}",
                "/api/profiles/**",
                "/api/recruitments/**",
                "/api/tags/**").permitAll()
            auth.requestMatchers("/", "/error", "/oauth2/**", "/api/tokens/refresh").permitAll()
            auth.anyRequest().authenticated()
        }
        .exceptionHandling { handler -> handler.authenticationEntryPoint(entryPoint) }
        .sessionManagement {
            it.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
        }
        .build()
}